scripts/Muon/GUI/Common/contexts/phase_table_context.py

```diff
@@ -23,18 +23,18 @@ def add_phase_table(self, workspace):
     def get_phase_table_list(self, instrument):
         return [phase_table.workspace_name for phase_table in self.phase_tables if instrument in phase_table.workspace_name]
 
-    def add_phase_quad(self, workspace):
-        self.phase_quad = [item for item in self.phase_quad if item.workspace_name != workspace.workspace_name]
-        self.phase_quad.append(workspace)
+    def add_phase_quad(self, workspace, run_list):
+        self.phase_quad = {key: item for key, item in self.phase_quad.items() if item.workspace_name != workspace.workspace_name}
+        self.phase_quad.update({run_list: workspace})
 
     def get_phase_quad(self, instrument, run):
-        return [phase_quad.workspace_name for phase_quad in self.phase_quad if instrument in phase_quad.workspace_name
-                and run in phase_quad.workspace_name]
+        return [phase_quad.workspace_name for key, phase_quad in self.phase_quad.items() if instrument in phase_quad.workspace_name
+                and run == key]
 
     def remove_workspace_by_name(self, workspace_name):
         self.phase_tables = [item for item in self.phase_tables if item.workspace_name != workspace_name]
-        self.phase_quad = [item for item in self.phase_tables if item.workspace_name != workspace_name]
+        self.phase_quad = {key: item for key, item in self.phase_quad.items() if item.workspace_name != workspace_name}
```

Comment:
instead of recreating the dict you could just use `del` https://stackoverflow.com/questions/5844672/delete-an-element-from-a-dictionary

Reply:
I'm not deleting items by key here but based on a property of the value. To use del I'd therefore have to iterate through the dictionary doing this check anyway. This leads me to iterating anyway and introduces the problems inherent in removing items from an iterable as you iterate through it. As the dictionaries are always going to be small I think the comprehension here is easier.